File: core/java/lib/src/main/java/com/microsoft/kiota/core/serialization/ParseNodeFactoryRegistry.java
package com.microsoft.kiota.core.serialization;

import java.io.InputStream;
import java.util.HashMap;
import java.util.Objects;

import javax.annotation.Nonnull;

import com.microsoft.kiota.serialization.ParseNode;
import com.microsoft.kiota.serialization.ParseNodeFactory;

public class ParseNodeFactoryRegistry implements ParseNodeFactory {
    public HashMap<String, ParseNodeFactory> contentTypeAssociatedFactories = new HashMap<>();
    @Override
    @Nonnull
    public ParseNode getParseNode(@Nonnull final String contentType, @Nonnull final InputStream rawResponse) {
        Objects.requireNonNull(contentType, "parameter contentType cannot be null");
        Objects.requireNonNull(rawResponse, "parameter rawResponse cannot be null");
        if(contentType.isEmpty()) {
            throw new NullPointerException("contentType cannot be empty");
        }
        if(contentTypeAssociatedFactories.containsKey(contentType)) {
            return contentTypeAssociatedFactories.get(contentType).getParseNode(contentType, rawResponse);
        } else {
            throw new RuntimeException("Content type " + contentType + " does not have a factory to be parsed");
        }
    }
}

File: abstractions/ruby/microsoft_kiota_abstractions/lib/microsoft_kiota_abstractions/serialization/parse_node_factory_registry.rb
require_relative 'parse_node_factory'

module MicrosoftKiotaAbstractions
  class ParseNodeFactoryRegistry
    include ParseNodeFactory

    class << self
    attr_accessor :default_instance
    def default_instance; @default_instance ||= ParseNodeFactoryRegistry.new; end
    end

    def default_instance
    self.class.default_instance
    end

    def content_type_associated_factories
      @content_type_associated_factories ||= Hash.new
    end

    def get_parse_node(content_type, content)
      if !content_type
        raise Exception.new 'content type cannot be undefined or empty'
      end
      if !content
        raise Exception.new 'content cannot be undefined or empty'
      end
      vendor_specific_content_type = content_type.split(';').first
      factory = @content_type_associated_factories[vendor_specific_content_type]
      if factory
        return factory.get_parse_node(vendor_specific_content_type, content)
      end

      clean_content_type = vendor_specific_content_type.gsub(/[^\/]+\+/i, '')
      factory = @content_type_associated_factories[clean_content_type]
      if factory
        return factory.get_parse_node(clean_content_type, content)
      end
      raise Exception.new "Content type #{contentType} does not have a factory to be parsed"
    end
        
  end
end
